op-chain-ops/cmd/check-derivation/main.go

```diff
@@ -0,0 +1,432 @@
+	prevHash, ok := blockMap[number]
+	if ok {
+		if prevHash != hash {
+			log.Error("Unsafe head reorg", "blockNum:", number,
+				"prevHash", prevHash.String(), "currHash", hash.String())
+		}
+	}
+}
+
+// detectL2Reorg polls safe heads and detects l2 unsafe block reorg.
+func detectL2Reorg(cliCtx *cli.Context) error {
+	ctx := context.Background()
+	clients, err := newClientsFromContext(ctx, cliCtx)
+	if err != nil {
+		return err
+	}
+	var pollingInterval = time.Duration(cliCtx.Int("polling-interval")) * time.Millisecond
+	// blockMap maps blockNumber to blockHash
+	blockMap := make(map[uint64]common.Hash)
+	var prevUnsafeHeadNum uint64
+	for {
+		unsafeHeadBlockId, parentHash, err := getUnsafeHead(ctx, clients.RpcClient)
+		if err != nil {
+			return fmt.Errorf("failed to fetch unsafe head: %w", err)
+		}
+		checkReorg(blockMap, unsafeHeadBlockId.Number-1, parentHash)
+		checkReorg(blockMap, unsafeHeadBlockId.Number, unsafeHeadBlockId.Hash)
+
+		if unsafeHeadBlockId.Number > prevUnsafeHeadNum {
+			log.Info("Fetched Unsafe block", "blockNum", unsafeHeadBlockId.Number, "hash", unsafeHeadBlockId.Hash.String())
+		}
+
+		blockMap[unsafeHeadBlockId.Number-1] = parentHash
+		blockMap[unsafeHeadBlockId.Number] = unsafeHeadBlockId.Hash
+		prevUnsafeHeadNum = unsafeHeadBlockId.Number
+		time.Sleep(pollingInterval)
+	}
+}
+
+// getRandomAddress returns vanity address of the form 0x000000000000000000000000[random 32 bits][prefix]
+// example: 0x00000000000000000000000030bd3402deadbeef
+func getRandomAddress(rng *rand.Rand, prefix uint64) common.Address {
+	var vanity uint64 = prefix + (uint64(rng.Uint32()) << 32)
+	return common.HexToAddress(fmt.Sprintf("0x%X", vanity))
+}
+
+func getPrivateKey(cliCtx *cli.Context) (*ecdsa.PrivateKey, error) {
+	privateKey, err := crypto.HexToECDSA(cliCtx.String("private-key"))
+	if err != nil {
+		return nil, fmt.Errorf("invalid private key: %w", err)
+	}
+	return privateKey, nil
+}
+
+func getSenderAddress(privateKey *ecdsa.PrivateKey) (common.Address, error) {
+	publicKey := privateKey.Public()
+	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
+	if !ok {
+		return common.Address{}, fmt.Errorf("error casting public key to ECDSA")
+	}
+	from := crypto.PubkeyToAddress(*publicKeyECDSA)
+	return from, nil
+}
+
+// getRandomSignedTransaction returns signed tx which sends 1 wei to random address
+func getRandomSignedTransaction(ctx context.Context, ethClient *ethclient.Client, rng *rand.Rand, from common.Address, privateKey *ecdsa.PrivateKey, chainId *big.Int, txType int, protected bool) (*types.Transaction, error) {
+	randomAddress := getRandomAddress(rng, 0xDEADBEEF)
+	amount := big.NewInt(1)
+	nonce, err := ethClient.PendingNonceAt(ctx, from)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get nonce: %w", err)
+	}
+
+	gasPrice, err := ethClient.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to suggest gas price: %w", err)
+	}
+	data := testutils.RandomData(rng, 10)
```

Comment:
Because this is random, 4 transactions isn't actually enough to cover the tx space. TBH I'd suggest that one invocation of this script simply submits 4 transactions, one of each type.

Reply:
yeah you're right. as you said, this script makes at least 4 random txs, but ensures one of each type.
`getRandomSignedTransaction` is being called for each tx types including unprotected legacy tx.